Python/errors.c

```diff
@@ -471,10 +475,25 @@ _PyErr_GetExcInfo(PyThreadState *tstate,
                   PyObject **p_type, PyObject **p_value, PyObject **p_traceback)
 {
     _PyErr_StackItem *exc_info = _PyErr_GetTopmostException(tstate);
-    *p_type = exc_info->exc_type;
+
     *p_value = exc_info->exc_value;
     *p_traceback = exc_info->exc_traceback;
 
+    assert(*p_value == NULL ||
+           *p_value == Py_None ||
+           PyExceptionInstance_Check(*p_value));
+
+    *p_type = (*p_value != NULL && PyExceptionInstance_Check(*p_value)) ?
+              PyExceptionInstance_Class(*p_value) :
+              Py_None;
+
+    if (exc_info->exc_type == NULL || exc_info->exc_type == Py_None) {
+        assert(*p_type == Py_None);
+    }
+    else {
+        assert(*p_type == exc_info->exc_type);
+    }
```

Comment:
The logic here is a bit hard to follow. Let me see if this summarizes it:
```py
typ, value, traceback = exc_info
assert nullish(value) or isinstance(value, BaseException)
typ = None if isinstance(value, BaseException) else type(value)
if nullish(exc_info[0]):
    assert typ is None
else:
    assert typ is exc_info[0]
return typ, value, traceback
```
Maybe it would be easier to follow if you rewrote it like this?
```py
typ, value, traceback = exc_info
if nullish(value):
    assert nullish(type)
    typ = None
elif isinstance(value, BaseException):
    assert typ is type(value)
    typ = type(value)  # In case assertions are off and it was a lie
else:
    assert False
return typ, value, traceback
```
Which avoids referring back to `exc_info` once we've extracted everything from it. (I think the reason I found it hard to follow is that it refers back to `exc_info`.)

Also note this effectively checks
```py
assert nullish(type) == nullish(value)
```
Which is an invariant you also assert on lines 614-615 (in `_PyErr_ChainStackItem()`).

(Note that my head is still spinning and I'm still not 100% sure that that's equivalent, but I am pretty sure it is.)